# Group Lab 2 

## Cleaning & Preparing Data for Future Analysis

In [1]:
#Run this
import pandas as pd                    # imports pandas and calls the imported version 'pd'
import matplotlib.pyplot as plt        # imports the package and calls it 'plt'
import seaborn as sns                  # imports the seaborn package with the imported name 'sns'
sns.set()  

### Steam Data

Steam is the world's most popular PC Gaming hub. With a massive collection that includes everything from AAA blockbusters to small indie titles, great discovery tools can be super valuable for Steam. What can we learn about Steam gaming behavior?

The dataset we will be analyzing is comprised of a random sample of 500 steam users and their game play and purchase behaviors. It has the following columns:
* user_id
* game_name,
* activity:
    - purchase: indicating that the user has *purchased* the corresponding game
    - play: indicating that the user has *played* the corresponding game (for at least some amount of time.) 
* hours_played_if_play:
    - if the row corresponds to a 'play' activity, this number represents the number of hours the user has played the game
    - if the row corresponds to a 'purchase' activity, this number is always a 1 (and means nothing... it's a placeholder).

Observational units in this data frame are defined by the user-game combination.

For this lab, you will be preparing the dataset for future analysis.  To do this, you will need to clean the data as requested below.  

### 1. Reading in the Data

First, read the steam_sample.csv file into a dataframe.  Display the first five rows and the number of observations in the data.

In [2]:
df = pd.read_csv('steam_sample.csv')

In [3]:
df.head()

user_id  game_name  activity hours_played_if_play
0  308653033   Unturned  purchase                    1
1  308653033   Unturned      play                  0.6
2  308653033  theHunter  purchase                    1
3  144004384     Dota 2  purchase                    1
4  144004384     Dota 2      play                   22

In [12]:
len(df)

7806

### 2. Prepare the Data for Analysis

Suppose you are an intern at Steam.  Your supervisor would like to analyze the data that you have but knows that its current format is not ideal for the questions that she'd like to answer.  Therefore, she asks you to prepare the data for her.  She gives you a list of the criteria that she would like for the final data to have.  Complete each of the following data preparation and analysis tasks:

1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**, whether detectable by default or not.  Make sure that Python reads these missing values correctly.  Then, drop any observations with missing values.
2. Filter the data to include only the user-game combinations that have been **played** for at least 1 hour.
3. We'd like to exclude any super-users of Steam, as some of these users are not representative of a typical Steam user.  Exclude the **users** that have the 5 largest user-game play times.  That is, find the users with the 5 largest play times for a single game, and then exclude their data for all games.
4. Identify the most popular game based on the number of users who have **purchased** that game.  Filter the data to include only that game.
5. Finally, order the data by the number of game hours played for each user-game combination.  Identify the fourth highest amount of time spent playing the game, the median time that a user spent playing that game, and the total number of hours that users in our data set spent playing the game.

#1

In [5]:
df.isna().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    0
dtype: int64

In [6]:
df.dtypes

user_id                  int64
game_name               object
activity                object
hours_played_if_play    object
dtype: object

In [7]:
df['hours_played_if_play'].unique()

array(['1', '0.6', '22', '1028', '1008', '148', '108', '72', '36', '35',
       '32', '21', '16', '15.8', '8.6', '7.8', '7.3', '3.1', '1.9', '1.7',
       '1.1', '0.4', '153', '63', '26', '1.4', '639', '479', '70', '65',
       '33', '30', '19.8', '16.2', '11.3', '4.2', '3.9', '2.3', '0.8',
       '0.7', '0.5', '0.3', '396', '227', '13.4', '12.6', '11.2', '10.1',
       '2.4', '210', '1.2', '0.2', '13.2', '48', '110', '0.1', '67',
       '429', '5.5', '61', '1.6', '18.3', '9.9', '4.7', '1714', '441',
       '197', '147', '117', '86', '73', '49', '46', '31', '24', '20',
       '19.7', '18.2', '14.2', '11.6', '9.8', '9.7', '8.4', '6.5', '4.8',
       '4.1', '3.7', '2.9', '2.7', '2.1', '222', 'unknown', '14.9',
       '14.1', '83', '11.1', '3.2', '6.9', '395', '251', '9.3', '7.4',
       '54', '34', '1.8', '99', '98', '96', '29', '27', '23', '19.1',
       '18.7', '17.5', '17', '16.6', '14.7', '13.3', '10.7', '10.2', '10',
       '9.6', '9.5', '8', '7.9', '7.6', '7', '6.6', '6', '5.8', '5

In [16]:
dfc = pd.read_csv('steam_sample.csv', na_values ='unknown')
print(len(dfc))
dfc.head()

7806


user_id  game_name  activity  hours_played_if_play
0  308653033   Unturned  purchase                   1.0
1  308653033   Unturned      play                   0.6
2  308653033  theHunter  purchase                   1.0
3  144004384     Dota 2  purchase                   1.0
4  144004384     Dota 2      play                  22.0

In [31]:
dfc['hours_played_if_play'].unique()

array([1.0000e+00, 6.0000e-01, 2.2000e+01, 1.0280e+03, 1.0080e+03,
       1.4800e+02, 1.0800e+02, 7.2000e+01, 3.6000e+01, 3.5000e+01,
       3.2000e+01, 2.1000e+01, 1.6000e+01, 1.5800e+01, 8.6000e+00,
       7.8000e+00, 7.3000e+00, 3.1000e+00, 1.9000e+00, 1.7000e+00,
       1.1000e+00, 4.0000e-01, 1.5300e+02, 6.3000e+01, 2.6000e+01,
       1.4000e+00, 6.3900e+02, 4.7900e+02, 7.0000e+01, 6.5000e+01,
       3.3000e+01, 3.0000e+01, 1.9800e+01, 1.6200e+01, 1.1300e+01,
       4.2000e+00, 3.9000e+00, 2.3000e+00, 8.0000e-01, 7.0000e-01,
       5.0000e-01, 3.0000e-01, 3.9600e+02, 2.2700e+02, 1.3400e+01,
       1.2600e+01, 1.1200e+01, 1.0100e+01, 2.4000e+00, 2.1000e+02,
       1.2000e+00, 2.0000e-01, 1.3200e+01, 4.8000e+01, 1.1000e+02,
       1.0000e-01, 6.7000e+01, 4.2900e+02, 5.5000e+00, 6.1000e+01,
       1.6000e+00, 1.8300e+01, 9.9000e+00, 4.7000e+00, 1.7140e+03,
       4.4100e+02, 1.9700e+02, 1.4700e+02, 1.1700e+02, 8.6000e+01,
       7.3000e+01, 4.9000e+01, 4.6000e+01, 3.1000e+01, 2.4000e

#2

In [22]:
filtered = dfc[(dfc['hours_played_if_play']>=1) & (dfc['activity']=='play')]
filtered

user_id                        game_name activity  \
4     144004384                           Dota 2     play   
6      54103616  Counter-Strike Global Offensive     play   
8      54103616                   Counter-Strike     play   
10     54103616                      Left 4 Dead     play   
12     54103616                   The Sims(TM) 3     play   
...         ...                              ...      ...   
7789   99096740                 Assassin's Creed     play   
7791   99096740                  Darkest Dungeon     play   
7793   99096740                           Crysis     play   
7795   99096740              Assassin's Creed II     play   
7797   99096740               Hitman Blood Money     play   

      hours_played_if_play  
4                     22.0  
6                   1028.0  
8                   1008.0  
10                   148.0  
12                   108.0  
...                    ...  
7789                  10.5  
7791                   6.4  
7793                   5.3  
7795                   2.7  
7797                   1.3  

[2222 rows x 4 columns]

In [23]:
len(filtered)

2222

#3

In [25]:
newdf = filtered.sort_values(by='hours_played_if_play', na_position = 'first', ignore_index = True)
newdf

user_id                        game_name activity  \
0     100719181     FreeStyle2 Street Basketball     play   
1     118852041                       Dirty Bomb     play   
2     118852041                  Age of Chivalry     play   
3     161623464                     sZone-Online     play   
4     106147454                       Dishonored     play   
...         ...                              ...      ...   
2217   48798067            Mount & Blade Warband     play   
2218   99077905                           Dota 2     play   
2219   43684632  Counter-Strike Global Offensive     play   
2220   32749624                   Counter-Strike     play   
2221   73017395       Sid Meier's Civilization V     play   

      hours_played_if_play  
0                      1.0  
1                      1.0  
2                      1.0  
3                      1.0  
4                      1.0  
...                    ...  
2217                3178.0  
2218                3309.0  
2219                3626.0  
2220                4814.0  
2221               11754.0  

[2222 rows x 4 columns]

In [26]:
exclude_top5 = newdf[(newdf['user_id'] != 48798067) & (newdf['user_id'] != 99077905) & (newdf['user_id'] != 43684632)	
                     & (newdf['user_id'] != 32749624) & (newdf['user_id'] != 73017395)]
len(exclude_top5)

1833

#4

In [85]:
purchased = dfc[dfc['activity']=='purchase']
purchased['game_name'].value_counts()

Dota 2                             197
Team Fortress 2                     95
Counter-Strike Global Offensive     67
Unturned                            54
Counter-Strike Source               50
                                  ... 
Edna & Harvey The Breakout           1
Memoria                              1
The Dark Eye Chains of Satinav       1
The Night of the Rabbit              1
Ori and the Blind Forest             1
Name: game_name, Length: 1445, dtype: int64

In [144]:
len(purchased)

4931

In [87]:
most_pop = purchased[purchased['game_name']=='Dota 2']
most_pop

user_id game_name  activity  hours_played_if_play
3     144004384    Dota 2  purchase                   1.0
35     54103616    Dota 2  purchase                   1.0
67    239397807    Dota 2  purchase                   1.0
127   173857208    Dota 2  purchase                   1.0
147   219446813    Dota 2  purchase                   1.0
...         ...       ...       ...                   ...
7673  218668551    Dota 2  purchase                   1.0
7724  126690600    Dota 2  purchase                   1.0
7750  147444403    Dota 2  purchase                   1.0
7752  154531790    Dota 2  purchase                   1.0
7754   99096740    Dota 2  purchase                   1.0

[197 rows x 4 columns]

In [145]:
len(most_pop)

197

#5

In [148]:
df_5 = exclude_top5[exclude_top5['game_name']=='Dota 2']
print(len(df_5))
df_5

150


user_id game_name activity  hours_played_if_play
11    149081421    Dota 2     play                   1.0
20    210018093    Dota 2     play                   1.0
81    204412693    Dota 2     play                   1.1
116   165907900    Dota 2     play                   1.2
122   227153589    Dota 2     play                   1.2
...         ...       ...      ...                   ...
2208  212915541    Dota 2     play                1605.0
2209   99096740    Dota 2     play                1704.0
2210   87071236    Dota 2     play                1714.0
2212   93290614    Dota 2     play                1760.0
2215  131217195    Dota 2     play                2179.0

[150 rows x 4 columns]

In [135]:
df_5.hours_played_if_play.nlargest(4).iloc[[-1]]

2209    1704.0
Name: hours_played_if_play, dtype: float64

 The fourth highest amount of time spent playing the game is 1704 hours. 

In [93]:
df_5['hours_played_if_play'].median()

13.55

The median time that a user spent playing that game is 13.55hours. 

In [94]:
df_5['hours_played_if_play'].sum()

28634.0

Total time that a user spent playing that game is 28634 hours. 

### 3. A Supervisor Misunderstanding? 

As an intern, you aren't quite sure if you've understood your supervisor's instructions correctly.  Your supervisor is now on vacation, so you can't reach out for clarification immediately.  You find yourself wondering "have I prepared the data properly?"

As one way to check if you have followed her intended instructions, you decide to try to prepare the data by completing the tasks listed above in a different order.

For this question, you should complete the same preparation tasks, but this time in a different order.  Again, after each task has been completed, print the number of rows in the data.

1. Identify any values (if any) that have been encoded in the csv to represent a **missing value**, whether detectable by default or not.  Make sure that Python reads these missing values correctly.  
2. Identify the most second popular game based on the number of users who have **purchased** that game.
3. Filter the data to include only the number of hours **played** for the game identified in step 2.
4. Drop any observations in the dataset from Step 3 that have missing values.
4. We'd like to exclude any super-users of Steam, as some of these users are not representative of a typical Steam user.  Exclude the **users** in our current data that have the 5 largest user-game play times.  That is, find the users with the 5 largest play times for a single game, and then exclude all of their data.
5. Filter the data to include only the games that have been played for at least 1 hour.
6. Finally, order the data by the number of game hours played for each user-game combination.  Identify the fourth highest amount of time spent playing the game, the median time that a user spent playing that game, and the total number of hours that users in our dataset spent playing the game.

Do you have the same data and results as task 2?

#1

In [28]:
df_2 = pd.read_csv('steam_sample.csv')
df_2.isna().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    0
dtype: int64

In [29]:
df_2.dtypes

user_id                  int64
game_name               object
activity                object
hours_played_if_play    object
dtype: object

In [30]:
df_2['hours_played_if_play'].unique()

array(['1', '0.6', '22', '1028', '1008', '148', '108', '72', '36', '35',
       '32', '21', '16', '15.8', '8.6', '7.8', '7.3', '3.1', '1.9', '1.7',
       '1.1', '0.4', '153', '63', '26', '1.4', '639', '479', '70', '65',
       '33', '30', '19.8', '16.2', '11.3', '4.2', '3.9', '2.3', '0.8',
       '0.7', '0.5', '0.3', '396', '227', '13.4', '12.6', '11.2', '10.1',
       '2.4', '210', '1.2', '0.2', '13.2', '48', '110', '0.1', '67',
       '429', '5.5', '61', '1.6', '18.3', '9.9', '4.7', '1714', '441',
       '197', '147', '117', '86', '73', '49', '46', '31', '24', '20',
       '19.7', '18.2', '14.2', '11.6', '9.8', '9.7', '8.4', '6.5', '4.8',
       '4.1', '3.7', '2.9', '2.7', '2.1', '222', 'unknown', '14.9',
       '14.1', '83', '11.1', '3.2', '6.9', '395', '251', '9.3', '7.4',
       '54', '34', '1.8', '99', '98', '96', '29', '27', '23', '19.1',
       '18.7', '17.5', '17', '16.6', '14.7', '13.3', '10.7', '10.2', '10',
       '9.6', '9.5', '8', '7.9', '7.6', '7', '6.6', '6', '5.8', '5

#2

In [32]:
purchased_2 = df_2[df_2['activity']=='purchase']
print(len(purchased_2))
purchased_2['game_name'].value_counts()

4931


Dota 2                             197
Team Fortress 2                     95
Counter-Strike Global Offensive     67
Unturned                            54
Counter-Strike Source               50
                                  ... 
Edna & Harvey The Breakout           1
Memoria                              1
The Dark Eye Chains of Satinav       1
The Night of the Rabbit              1
Ori and the Blind Forest             1
Name: game_name, Length: 1445, dtype: int64

The most popular game is Dota 2. 

#3

In [34]:
filtered_2 = dfc[(dfc['game_name']=='Dota 2') & (dfc['activity']=='play')]
print(len(filtered_2))
filtered_2

197


user_id game_name activity  hours_played_if_play
4     144004384    Dota 2     play                  22.0
36     54103616    Dota 2     play                   3.1
68    239397807    Dota 2     play                   0.4
128   173857208    Dota 2     play                   0.6
148   219446813    Dota 2     play                 210.0
...         ...       ...      ...                   ...
7674  218668551    Dota 2     play                   2.2
7725  126690600    Dota 2     play                   0.2
7751  147444403    Dota 2     play                   0.7
7753  154531790    Dota 2     play                   0.1
7755   99096740    Dota 2     play                1704.0

[197 rows x 4 columns]

#4

In [36]:
filtered_2.isna().sum()

user_id                 0
game_name               0
activity                0
hours_played_if_play    0
dtype: int64

In [37]:
filtered_2.dtypes

user_id                   int64
game_name                object
activity                 object
hours_played_if_play    float64
dtype: object

In [42]:
len(filtered_2)

197

#5

In [44]:
filtered_2.sort_values(by='hours_played_if_play', na_position='first', ignore_index=True)

user_id game_name activity  hours_played_if_play
0    236191066    Dota 2     play                   0.1
1    141156585    Dota 2     play                   0.1
2    133231400    Dota 2     play                   0.1
3    165906966    Dota 2     play                   0.1
4    154531790    Dota 2     play                   0.1
..         ...       ...      ...                   ...
192   99096740    Dota 2     play                1704.0
193   87071236    Dota 2     play                1714.0
194   93290614    Dota 2     play                1760.0
195  131217195    Dota 2     play                2179.0
196   99077905    Dota 2     play                3309.0

[197 rows x 4 columns]

In [45]:
exclude_top5_2 = filtered_2[(filtered_2['user_id'] != 99077905) & (filtered_2['user_id'] != 131217195) & (filtered_2['user_id'] != 93290614)
                     & (filtered_2['user_id'] != 87071236) & (filtered_2['user_id'] != 99096740)]
print(len(exclude_top5_2))

192


#6

In [41]:
played_atleast_1hour = exclude_top5_2[exclude_top5_2['hours_played_if_play']>=1]
len(played_atleast_1hour)

147

In [128]:
played_atleast_1hour.sort_values(by='hours_played_if_play')

user_id game_name activity  hours_played_if_play
5392  210018093    Dota 2     play                   1.0
608   149081421    Dota 2     play                   1.0
5116  204412693    Dota 2     play                   1.1
3012  183315077    Dota 2     play                   1.2
152   227153589    Dota 2     play                   1.2
...         ...       ...      ...                   ...
6451  123917008    Dota 2     play                1301.0
5434  122620165    Dota 2     play                1337.0
2063  116831291    Dota 2     play                1343.0
3170   86866292    Dota 2     play                1540.0
3065  212915541    Dota 2     play                1605.0

[147 rows x 4 columns]

In [134]:
played_atleast_1hour.hours_played_if_play.nlargest(4).iloc[[-1]]

5434    1337.0
Name: hours_played_if_play, dtype: float64

In [130]:
played_atleast_1hour['hours_played_if_play'].median()

11.1

In [131]:
played_atleast_1hour['hours_played_if_play'].sum()

21279.4

The fourth highest amount of time spent playing the game is 1337 hours.

The median time that a user spent playing that game is 11.1 hours and the total number of hours that users in our dataset spent playing the game is 21279.4 hours.

The result of task 1 and task 2 is different. 

### 4. Match the description

You are a resourceful intern, so you keep searching for guidance on how best to clean the data.  You find an article that seems to have performed similar data processing steps.

The article describes their data processing as follows:

Only complete observations were considered for the second most popular game based on purchases.  Players who had played the game for 1 hour or more and were not in the top 5 players based on time were considered as the primary users for analysis.

Which of the two data preparations that you completed above most closely matches the article description?  Explain why you made this selection.

The first analysis dropped out the extreme players including the players who played the other games, not only 'Dota 2' (the most purchased game). 

On the contrary, the second analysis dropped out the extreme-time players only for the players of 'Dota 2'(the most purchased game). 

So, the second data preparation was the datat that matches the article discription.  